cli/command/swarm/progress/root_rotation.go

```diff
@@ -0,0 +1,171 @@
+package progress
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/streamformatter"
+	digest "github.com/opencontainers/go-digest"
+)
+
+const (
+	certsRotatedStr = "  nodes rotated TLS certificates"
+	rootsRotatedStr = "   nodes rotated CA certificates"
+)
+
+// RootRotationProgress outputs progress information for convergence of a root rotation.
+func RootRotationProgress(ctx context.Context, dclient client.APIClient, progressWriter io.WriteCloser) error {
+	defer progressWriter.Close()
+
+	progressOut := streamformatter.NewJSONStreamFormatter().NewProgressOutput(progressWriter, false)
+
+	sigint := make(chan os.Signal, 1)
+	signal.Notify(sigint, os.Interrupt)
+	defer signal.Stop(sigint)
+
+	var (
+		updater     *rootRotationProgressUpdater
+		converged   bool
+		convergedAt time.Time
+		monitor     = 3 * time.Second
+	)
+
+	for {
+		info, err := dclient.SwarmInspect(ctx)
+		if err != nil {
+			return err
+		}
+
+		if updater == nil {
+			updater = &rootRotationProgressUpdater{
+				progressOut: progressOut,
+			}
+		}
+
+		if converged && time.Since(convergedAt) >= monitor {
+			return nil
+		}
+
+		nodes, err := dclient.NodeList(ctx, types.NodeListOptions{})
+		if err != nil {
+			return err
+		}
+
+		updater.update(info.ClusterInfo.TLSInfo, nodes, info.ClusterInfo.RootRotationInProgress)
+		converged = updater.done
+		if converged {
+			if convergedAt.IsZero() {
+				convergedAt = time.Now()
+			}
+			wait := monitor - time.Since(convergedAt)
+			if wait >= 0 {
+				progressOut.WriteProgress(progress.Progress{
+					// Ideally this would have no ID, but
+					// the progress rendering code behaves
+					// poorly on an "action" with no ID. It
+					// returns the cursor to the beginning
+					// of the line, so the first character
+					// may be difficult to read. Then the
+					// output is overwritten by the shell
+					// prompt when the command finishes.
+					ID:     "verify",
+					Action: fmt.Sprintf("Waiting %d seconds to verify that the roots are stable...", wait/time.Second+1),
```

Comment:
@aaronlehmann It was only to cover another root rotation - agree it's probably not necessary though.